Antidote/AbstractCallViewController.h

```diff
@@ -8,31 +8,45 @@
 
 #import <UIKit/UIKit.h>
 
+#import "OCTToxAVConstants.h"
+
 @class OCTSubmanagerCalls;
 @class OCTChat;
 @class OCTCall;
 
+@protocol AbstractCallViewControllerDelegate <NSObject>
+
+@required
+- (void)dismissCurrentCall;
+- (void)callAccept;
+- (BOOL)sendCallControl:(OCTToxAVCallControl)control error:(NSError **)error;
+
+- (void)otherCallAccept:(BOOL)accept;
+
+@property (nonatomic, assign) BOOL enableMicrophone;
+
+@end
```

Comment:
I think that it would be nice if controllers would be as light as possible. There is no need for controllers to know something about call. Also there is no need for them to know about OCTToxAVCallControl. They will be responsible only for presenting text and handling button touches.

You can pass touches and handle them in CallsManager. Controllers will have properties which CallsManager will update. In setter of property you can update UI immediately. You don't have to recreate controller when changing active call - just update caller name and call duration.

 Also I don't think that you should move everything to Abstract controller. You can leave it just for blurred background.

So API as I see it is following. I may be missing something, but hope you get general idea:

```
@class DialingCallViewController
@property (weak, nonatomic) id<DialingCallViewControllerDelegate> delegate;
@property (strong, nonatomic) NSString *callerName;
@end

@protocol DialingCallViewControllerDelegate
- (void)dialingCallDeclineButtonPressed;
@end
```

```
@class RingingCallViewController
@property (weak, nonatomic) id<RingingCallViewControllerDelegate> delegate;
@property (strong, nonatomic) NSString *callerName;
@end

@protocol RingingCallViewControllerDelegate
- (void)ringingCallAnswerButtonPressed;
- (void)ringingCallDeclineButtonPressed;
@end
```

```
@class ActiveCallViewController
@property (weak, nonatomic) id<AbstractCallViewControllerDelegate> delegate;

@property (strong, nonatomic) NSString *callerName;
@property (assign, nonatomic) NSTimeInterval callDuration;
@property (assign, nonatomic) BOOL muteSelected;
@property (assign, nonatomic) BOOL speakerSelected;

@property (assign, nonatomic) BOOL showIncomingCallView;
@property (strong, nonatomic) NSString *incomingCallCallerName;

// An array with "paused calls" objects. Each object has following properties:
// NSString *callerName;
// NSTimeInterval startHoldInterval; // from that time you can calculate on hold duration
@property (strong, nonatomic) NSArray *pausedCallsArray;
@end

@protocol ActiveCallViewControllerDelegate
- (void)activeCallDeclineButtonPressed;
- (void)activeCallMuteButtonPressed;
- (void)activeCallSpeakerButtonPressed;
- (void)activeCallDeclineIncomingCallButtonPressed;
- (void)activeCallAnswerIncomingCallButtonPressed;
- (void)activeCallPausedCallSelectedAtIndex:(NSUInteger)index;
@end
```